### PR TITLE
Timeline data structure

### PR DESCRIPTION
**Pull request by LorenzoLotti**
A collection of System.DateTime and values sorted by the System.DateTime field.

Please include a summary of the change. Please also include relevant motivation and context (if applicable).

- [YES] I have performed a self-review of my code
- [YES] My code follows the style guidelines of this project
- [NO HARD AREAS IN THIS CODE] I have added comments to hard-to-understand areas of my code
- [YES] I have added tests that prove my fix is effective or that my feature works
- [YES] New and existing unit tests pass locally with my changes
- [?] I have made corresponding changes to the comments
- [NO ] I have made corresponding changes to the README.md

### CODE REVIEW

File: Timeline.cs [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>, IDisposable
```
Comment (siriak) [inline Timeline.cs:481]: Codacy Issue found: ['IEnumerator<(DateTime Time, TValue Value)>' implements 'IDisposable' so 'IDisposable' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- Timeline.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: Timeline.cs [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+    /// <summary>
+    /// The count of items.
+    /// </summary>
+    public int Count { get => timeline.Count; }
+
+    /// <summary>
+    /// Get all values associated with <paramref name="time"/>.
+    /// </summary>
+    /// <returns>Values associated with <paramref name="time"/>.</returns>
+    public TValue[] this[DateTime time]
```
Comment (siriak) [inline Timeline.cs:28]: Codacy Issue found: [Use string or an integral type here, or refactor this indexer into a method.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs (hunk 1 above)
Comment (siriak) [inline Timeline.cs:481]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+/// <summary>
+/// A collection of <see cref="System.DateTime"/> and <see cref="TValue"/> sorted by <see cref="System.DateTime"/> field.
+/// </summary>
+/// <typeparam name="TValue">Value associated with a <see cref="System.DateTime"/>.</typeparam>
+internal class Timeline<TValue>:
+    IEnumerable<(DateTime Time, TValue Value)>,
```
Comment (siriak) [inline Timeline.cs:11]: Codacy Issue found: ['ICollection<(DateTime Time, TValue Value)>' implements 'IEnumerable<(DateTime Time, TValue Value)>' so 'IEnumerable<(DateTime Time, TValue Value)>' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>, IDisposable
+    {
+        internal int index;
+        internal Timeline<TValue> timeline;
+        object IEnumerator.Current { get => timeline.timeline[index]; }
+        (DateTime Time, TValue Value) IEnumerator<(DateTime Time, TValue Value)>.Current { get => timeline.timeline[index]; }
+        void IDisposable.Dispose() { }
```
Comment (siriak) [inline Timeline.cs:487]: Codacy Issue found: [Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- Timeline.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: Timeline.cs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>
```
Comment (siriak) [inline Timeline.cs:482]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+    /// <summary>
+    /// The count of items.
+    /// </summary>
+    public int Count { get => timeline.Count; }
+
+    /// <summary>
+    /// Get all values associated with <paramref name="time"/>.
+    /// </summary>
+    /// <returns>Values associated with <paramref name="time"/>.</returns>
+    public TValue[] this[DateTime time]
```
Comment (siriak) [inline Timeline.cs:29]: Codacy Issue found: [Use string or an integral type here, or refactor this indexer into a method.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+    /// <param name="other">Another <see cref="Timeline{TValue}"/> object.</param>
+    public bool Equals(Timeline<TValue> other)
+    {
+        return this == other;
+    }
+
+    /// <summary>
+    /// Returns a <see cref="System.Object"/> clone of the current instance.
+    /// </summary>
+    public object Clone()
```
Comment (siriak) [inline Timeline.cs:100]: Could you remove cloning? I believe it's not essential from educational standpoint and complicates the overall design.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- Timeline.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: Timeline.cs [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>
+    {
+        private int index;
+        private Timeline<TValue> timeline;
```
Comment (siriak) [inline Timeline.cs:476]: Codacy Issue found: [Make 'timeline' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: Timeline.cs [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>
```
Comment (siriak) [inline Timeline.cs:473]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

Comment (siriak) [inline Timeline.cs:473]: Codacy Issue found: [This class can't be instantiated; make its constructor 'public'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Timeline.cs  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: Timeline.cs (hunk 9 above)
Comment (siriak) [inline Timeline.cs:473]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  124
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+
+    /// <summary>
+    /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+    /// </summary>
+    public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+    {
+        return !(left == right);
+    }
+
+    private class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>
```
Comment (siriak) [inline DataStructures/Timeline.cs:473]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: DataStructures.Tests/TimelineTests.cs [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+            foreach (var pair in timeline)
+                timelineList.Add(pair);
+            var dictionaryList = new System.Collections.Generic.List<(DateTime Time, string Value)>();
+            foreach (var (key, value) in dictionary)
+                dictionaryList.Add((key, value));
+            timelineList.OrderBy(pair => pair.Time);
+            dictionaryList.OrderBy(pair => pair.Time);
+            Assert.IsTrue(timelineList.Count == dictionaryList.Count);
+            for (int i = 0; i < timelineList.Count; i++)
+                Assert.IsTrue(timelineList[i].Time == dictionaryList[i].Time && timelineList[i].Value == timelineList[i].Value);
```
Comment (siriak) [inline DataStructures.Tests/TimelineTests.cs:579]: Codacy Issue found: [Correct one of the identical expressions on both sides of operator '=='.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  124
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs (hunk 10 above)
Comment (siriak) [inline DataStructures/Timeline.cs:473]: Codacy Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  124
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+                    if (timeline[i].Time == time)
+                    {
+                        if (c == value.Length)
+                            timeline.RemoveAt(i);
+                        else
+                            timeline[i] = (timeline[i].Time, value[++c]);
+                    }
+                }
+                while (i < c)
+                    Add(time, value[c]);
```
Comment (siriak) [inline DataStructures/Timeline.cs:57]: Could you comment this piece of code a bit because I don't completely get an idea of what's going on

**Comment by LorenzoLotti** [comment]
Of course

Il ven 19 giu 2020, 14:30 Andrii Siriak <notifications@github.com> ha
scritto:

File: DataStructures/Timeline.cs [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+using System.Linq;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// A collection of <see cref="DateTime"/> and <see cref="TValue"/> sorted by <see cref="DateTime"/> field.
+    /// </summary>
+    /// <typeparam name="TValue">Value associated with a <see cref="DateTime"/>.</typeparam>
+    public class Timeline<TValue>:
+        IEnumerable<(DateTime Time, TValue Value)>,
```
Comment (siriak) [inline DataStructures/Timeline.cs:27]: Codacy found an issue: ['ICollection<(DateTime Time, TValue Value)>' implements 'IEnumerable<(DateTime Time, TValue Value)>' so 'IEnumerable<(DateTime Time, TValue Value)>' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures.Tests/TimelineTests.cs [hunk 14]
```diff
@@ -0,0 +1,725 @@
...
```
Comment (siriak) [inline DataStructures.Tests/TimelineTests.cs:676]: Codacy found an issue: [Correct one of the identical expressions on both sides of operator '=='.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: DataStructures/Timeline.cs [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+
+        /// <summary>
+        /// Returns true if <paramref name="left"/> is unequal to <paramref name="right"/>.
+        /// </summary>
+        public static bool operator !=(Timeline<TValue> left, Timeline<TValue> right)
+        {
+            return !(left == right);
+        }
+
+        private sealed class TimelineEnumerator: IEnumerator<(DateTime Time, TValue Value)>, IDisposable
```
Comment (siriak) [inline DataStructures/Timeline.cs:528]: Codacy found an issue: ['IEnumerator<(DateTime Time, TValue Value)>' implements 'IDisposable' so 'IDisposable' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs (hunk 13 above)
Comment (siriak) [inline DataStructures/Timeline.cs:27]: Codacy found an issue: ['ICollection<(DateTime Time, TValue Value)>' implements 'IEnumerable<(DateTime Time, TValue Value)>' so 'IEnumerable<(DateTime Time, TValue Value)>' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

File: DataStructures/Timeline.cs (hunk 15 above)
Comment (siriak) [inline DataStructures/Timeline.cs:528]: Codacy found an issue: ['IEnumerator<(DateTime Time, TValue Value)>' implements 'IDisposable' so 'IDisposable' can be removed from the inheritance list.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+        IEquatable<Timeline<TValue>>
+    {
+        private List<(DateTime Time, TValue Value)> timeline = new List<(DateTime Time, TValue Value)>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Timeline{TValue}"/> class.
+        /// </summary>
+        public Timeline()
+        {
+            // todo: improve performance and consider removing unnecessary methods
```
Comment (siriak) [inline DataStructures/Timeline.cs:37]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs (hunk 16 above)
Comment (siriak) [inline DataStructures/Timeline.cs:37]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by LorenzoLotti** [comment]
@siriak i think you can merge it now.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs [hunk 17]
```diff
@@ -0,0 +1,622 @@
...
```
Comment (siriak) [inline DataStructures/Timeline.cs:37]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures.Tests/TimelineTests.cs  4
- DataStructures/Timeline.cs  5
         

Clones added
============
- DataStructures.Tests/TimelineTests.cs  158
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777&bid=18276465)

File: DataStructures/Timeline.cs (hunk 17 above)
Comment (siriak) [inline DataStructures/Timeline.cs:37]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5574777)

**Pull request merged**